subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/AbstractFileWatcherUpdater.java

```diff
@@ -16,28 +16,125 @@
+        }
+        return newRoot;
     }
 
     @CheckReturnValue
     protected abstract SnapshotHierarchy doUpdateVfsOnBuildStarted(SnapshotHierarchy root);
+
+    @Override
+    public void virtualFileSystemContentsChanged(Collection<FileSystemLocationSnapshot> removedSnapshots, Collection<FileSystemLocationSnapshot> addedSnapshots, SnapshotHierarchy root) {
+        boolean contentsChanged = handleVirtualFileSystemContentsChanged(removedSnapshots, addedSnapshots, root);
+        if (contentsChanged) {
+            updateWatchedHierarchies(root);
+        }
+    }
+
+    protected abstract boolean handleVirtualFileSystemContentsChanged(Collection<FileSystemLocationSnapshot> removedSnapshots, Collection<FileSystemLocationSnapshot> addedSnapshots, SnapshotHierarchy root);
+
+    @Override
+    public SnapshotHierarchy updateVfsOnBuildFinished(SnapshotHierarchy root, WatchMode watchMode, int maximumNumberOfWatchedHierarchies) {
+        SnapshotHierarchy newRoot = watchableHierarchies.removeUnwatchableContentOnBuildFinished(
+            root,
+            watchMode,
+            watchedHierarchies::contains,
+            maximumNumberOfWatchedHierarchies,
+            createInvalidator()
+        );
+
+        if (root != newRoot) {
+            updateWatchedHierarchies(newRoot);
+        }
+        LOGGER.info("Watched directory hierarchies: {}", watchedHierarchies.getWatchedRoots());
+        return newRoot;
+    }
+
+    @Override
+    public Collection<File> getWatchedRoots() {
+        return watchedHierarchies.getWatchedRoots();
+    }
+
+    @Override
+    public void triggerWatchProbe(String path) {
+        watchableHierarchies.triggerWatchProbe(path);
+    }
+
+    protected abstract WatchableHierarchies.Invalidator createInvalidator();
+
+    private void updateWatchedHierarchies(SnapshotHierarchy root) {
+        Set<File> oldWatchedRoots = watchedHierarchies.getWatchedRoots();
+        watchedHierarchies = WatchedHierarchies.resolveWatchedHierarchies(watchableHierarchies, root);
+        Set<File> newWatchedRoots = watchedHierarchies.getWatchedRoots();
+
+        if (newWatchedRoots.isEmpty()) {
+            LOGGER.info("Not watching anything anymore");
+        }
+        Set<File> hierarchiesToStopWatching = new HashSet<>(oldWatchedRoots);
+        Set<File> hierarchiesToStartWatching = new HashSet<>(newWatchedRoots);
+        hierarchiesToStopWatching.removeAll(newWatchedRoots);
+        hierarchiesToStartWatching.removeAll(oldWatchedRoots);
+        if (hierarchiesToStartWatching.isEmpty() && hierarchiesToStopWatching.isEmpty()) {
+            return;
+        }
+        if (!hierarchiesToStopWatching.isEmpty()) {
+            stopWatchingHierarchies(hierarchiesToStopWatching);
+        }
+        if (!hierarchiesToStartWatching.isEmpty()) {
+            hierarchiesToStartWatching.forEach(locationToWatchValidator::validateLocationToWatch);
+            hierarchiesToStartWatching.forEach(watchableHierarchies::armWatchProbe);
```

Comment:
I think we need to arm the probe after start watching the new hierarchies, right?

Reply:
Good point.

Reply:
Changed it around.